File: app/src/main/java/com/bopr/android/smailer/remote/RemoteControlTaskParser.java
package com.bopr.android.smailer.remote;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Scanner;

import static com.bopr.android.smailer.remote.RemoteControlTask.ADD_PHONE_TO_BLACKLIST;
import static com.bopr.android.smailer.remote.RemoteControlTask.ADD_PHONE_TO_WHITELIST;
import static com.bopr.android.smailer.remote.RemoteControlTask.ADD_TEXT_TO_BLACKLIST;
import static com.bopr.android.smailer.remote.RemoteControlTask.ADD_TEXT_TO_WHITELIST;
import static com.bopr.android.smailer.remote.RemoteControlTask.REMOVE_PHONE_FROM_BLACKLIST;
import static com.bopr.android.smailer.remote.RemoteControlTask.REMOVE_PHONE_FROM_WHITELIST;
import static com.bopr.android.smailer.remote.RemoteControlTask.REMOVE_TEXT_FROM_BLACKLIST;
import static com.bopr.android.smailer.remote.RemoteControlTask.REMOVE_TEXT_FROM_WHITELIST;
import static com.bopr.android.smailer.util.AddressUtil.PHONE_PATTERN;
import static com.bopr.android.smailer.util.TextUtil.QUOTED_TEXT_PATTERN;

/**
 * Parses text into remote control task.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
class RemoteControlTaskParser {
    
    RemoteControlTask parse(String text) {
        RemoteControlTask task = new RemoteControlTask();
        Scanner scanner = new Scanner(text).useDelimiter("\\W+");

        if (hasNextToken(scanner, "(?i:DEVICE)")) {
            task.setAcceptor(nextQuoted(scanner));

            switch (nextToken(scanner, "(?i:ADD|REMOVE|SEND)")) {
                case "SEND":
                    if (hasNextToken(scanner, "(?i:SMS)")) {
                        task.setAction(RemoteControlTask.SEND_SMS_TO_CALLER);
                        task.setArgument("text", nextQuoted(scanner));
                        task.setArgument("phone", nextPhone(scanner));
                    }
                    break;
                case "ADD":
                    switch (nextToken(scanner, "(?i:PHONE|TEXT)")) {
                        case "PHONE":
                            task.setArgument(nextPhone(scanner));
                            switch (nextToken(scanner, "(?i:BLACKLIST|WHITELIST)")) {
                                case "BLACKLIST":
                                    task.setAction(ADD_PHONE_TO_BLACKLIST);
                                    break;
                                case "WHITELIST":
                                    task.setAction(ADD_PHONE_TO_WHITELIST);
                                    break;
                            }
                            break;
                        case "TEXT":
                            task.setArgument(nextQuoted(scanner));
                            switch (nextToken(scanner, "(?i:BLACKLIST|WHITELIST)")) {
                                case "BLACKLIST":
                                    task.setAction(ADD_TEXT_TO_BLACKLIST);
                                    break;
                                case "WHITELIST":
                                    task.setAction(ADD_TEXT_TO_WHITELIST);
                                    break;
                            }
                            break;
                    }
                case "REMOVE":
                    switch (nextToken(scanner, "(?i:PHONE|TEXT)")) {
                        case "PHONE":
                            task.setArgument(nextPhone(scanner));
                            switch (nextToken(scanner, "(?i:BLACKLIST|WHITELIST)")) {
                                case "BLACKLIST":
                                    task.setAction(REMOVE_PHONE_FROM_BLACKLIST);
                                    break;
                                case "WHITELIST":
                                    task.setAction(REMOVE_PHONE_FROM_WHITELIST);
                                    break;
                            }
                            break;
                        case "TEXT":
                            task.setArgument(nextQuoted(scanner));
                            switch (nextToken(scanner, "(?i:BLACKLIST|WHITELIST)")) {
                                case "BLACKLIST":
                                    task.setAction(REMOVE_TEXT_FROM_BLACKLIST);
                                    break;
                                case "WHITELIST":
                                    task.setAction(REMOVE_TEXT_FROM_WHITELIST);
                                    break;
                            }
                            break;
                    }
                    break;
            }
        }

        return task;
    }

    private boolean hasNextToken(Scanner scanner, String pattern) {
        return scanner.findWithinHorizon(pattern, 0) != null;
    }

    @NonNull
    private String nextToken(Scanner scanner, String pattern) {
        String s = scanner.findWithinHorizon(pattern, 0);
        return s != null ? s.toUpperCase() : "";
    }

    @Nullable
    private String nextQuoted(Scanner scanner) {
        if (scanner.findWithinHorizon(QUOTED_TEXT_PATTERN, 0) != null) {
            return scanner.match().group(1);
        }
        return null;
    }

    @Nullable
    private String nextPhone(Scanner scanner) {
        return scanner.findWithinHorizon(PHONE_PATTERN, 0);
    }
}

File: app/src/main/kotlin/com/bopr/android/smailer/remote/RemoteControlTaskParser.kt
package com.bopr.android.smailer.remote

import com.bopr.android.smailer.remote.RemoteControlTask.Companion.ADD_PHONE_TO_BLACKLIST
import com.bopr.android.smailer.remote.RemoteControlTask.Companion.ADD_PHONE_TO_WHITELIST
import com.bopr.android.smailer.remote.RemoteControlTask.Companion.ADD_TEXT_TO_BLACKLIST
import com.bopr.android.smailer.remote.RemoteControlTask.Companion.ADD_TEXT_TO_WHITELIST
import com.bopr.android.smailer.remote.RemoteControlTask.Companion.REMOVE_PHONE_FROM_BLACKLIST
import com.bopr.android.smailer.remote.RemoteControlTask.Companion.REMOVE_PHONE_FROM_WHITELIST
import com.bopr.android.smailer.remote.RemoteControlTask.Companion.REMOVE_TEXT_FROM_BLACKLIST
import com.bopr.android.smailer.remote.RemoteControlTask.Companion.REMOVE_TEXT_FROM_WHITELIST
import com.bopr.android.smailer.remote.RemoteControlTask.Companion.SEND_SMS_TO_CALLER
import com.bopr.android.smailer.util.PHONE_REGEX
import com.bopr.android.smailer.util.QUOTATION_PATTERN
import com.bopr.android.smailer.util.QUOTATION_REGEX
import java.util.Locale
import java.util.Scanner
import java.util.regex.Pattern

/**
 * Parses text into remote control task.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
internal class RemoteControlTaskParser {

    fun parse(text: String): RemoteControlTask? {
        val scanner = Scanner(text).useDelimiter("\\W+")
        if (hasNextToken(scanner, "(?i:DEVICE)")) {
            val task = RemoteControlTask(nextQuoted(scanner))
            when (nextToken(scanner, "(?i:ADD|PUT|REMOVE|DELETE|SEND)")) {
                "SEND" ->
                    if (hasNextToken(scanner, "(?i:SMS)")) {
                        task.action = SEND_SMS_TO_CALLER
                        task.arguments["text"] = nextQuoted(scanner)
                        task.arguments["phone"] = nextPhone(scanner)
                    }

                "ADD", "PUT" -> {
                    when (nextToken(scanner, "(?i:PHONE|TEXT)")) {
                        "PHONE" -> {
                            task.argument = nextPhone(scanner)
                            when (nextToken(scanner, "(?i:BLACKLIST|WHITELIST)")) {
                                "BLACKLIST" -> task.action = ADD_PHONE_TO_BLACKLIST
                                "WHITELIST" -> task.action = ADD_PHONE_TO_WHITELIST
                            }
                        }

                        "TEXT" -> {
                            task.argument = nextQuoted(scanner)
                            when (nextToken(scanner, "(?i:BLACKLIST|WHITELIST)")) {
                                "BLACKLIST" -> task.action = ADD_TEXT_TO_BLACKLIST
                                "WHITELIST" -> task.action = ADD_TEXT_TO_WHITELIST
                            }
                        }
                    }
                    when (nextToken(scanner, "(?i:PHONE|TEXT)")) {
                        "PHONE" -> {
                            task.argument = nextPhone(scanner)
                            when (nextToken(scanner, "(?i:BLACKLIST|WHITELIST)")) {
                                "BLACKLIST" -> task.action = REMOVE_PHONE_FROM_BLACKLIST
                                "WHITELIST" -> task.action = REMOVE_PHONE_FROM_WHITELIST
                            }
                        }

                        "TEXT" -> {
                            task.argument = nextQuoted(scanner)
                            when (nextToken(scanner, "(?i:BLACKLIST|WHITELIST)")) {
                                "BLACKLIST" -> task.action = REMOVE_TEXT_FROM_BLACKLIST
                                "WHITELIST" -> task.action = REMOVE_TEXT_FROM_WHITELIST
                            }
                        }
                    }
                }

                "REMOVE", "DELETE" -> when (nextToken(scanner, "(?i:PHONE|TEXT)")) {
                    "PHONE" -> {
                        task.argument = nextPhone(scanner)
                        when (nextToken(scanner, "(?i:BLACKLIST|WHITELIST)")) {
                            "BLACKLIST" -> task.action = REMOVE_PHONE_FROM_BLACKLIST
                            "WHITELIST" -> task.action = REMOVE_PHONE_FROM_WHITELIST
                        }
                    }

                    "TEXT" -> {
                        task.argument = nextQuoted(scanner)
                        when (nextToken(scanner, "(?i:BLACKLIST|WHITELIST)")) {
                            "BLACKLIST" -> task.action = REMOVE_TEXT_FROM_BLACKLIST
                            "WHITELIST" -> task.action = REMOVE_TEXT_FROM_WHITELIST
                        }
                    }
                }
            }
            return task
        }
        return null
    }

    private fun hasNextToken(scanner: Scanner, pattern: String): Boolean {
        return scanner.findWithinHorizon(pattern, 0) != null
    }

    private fun nextToken(scanner: Scanner, pattern: String): String {
        return scanner.findWithinHorizon(pattern, 0).uppercase(Locale.ROOT)
    }

    private fun nextQuoted(scanner: Scanner): String? {
        return scanner.findWithinHorizon(QUOTATION_PATTERN, 0)?.let {
            scanner.match().group(1)
        }
    }

    private fun nextPhone(scanner: Scanner): String? {
        return scanner.findWithinHorizon(PHONE_OR_QUOTATION_PATTERN, 0)?.let { text ->
            scanner.match().group(2)    /* found quotation */
                ?: text               /* found pure number */
        }
    }

    companion object {

        private val PHONE_OR_QUOTATION_PATTERN: Pattern =
            Pattern.compile("($QUOTATION_REGEX|$PHONE_REGEX)")
    }
}